### DISCUSSION TITLE
Incorrect State for operation: "running", allowed States: "[new,stopped]"

### DISCUSSION THREAD

**Discussion by iyosayi** [typegoose/mongodb-memory-server#493] [Q&A] [answered]
I am trying to run tests in my CI process using Github actions and mocha, but the tests don't run at all because I always get this error message.
I have checked google for a solution, but haven't come across any. Please how do I fix this? Because this is new to me.

```
Error: Incorrect State for operation: "running", allowed States: "[new,stopped]"
      at MongoMemoryServer.<anonymous> (node_modules/mongodb-memory-server-core/src/MongoMemoryServer.ts:269:15)
      at Generator.next (<anonymous>)
      at /home/runner/work/personal-project/personal-project/Backend/node_modules/tslib/tslib.js:117:75
      at new Promise (<anonymous>)
      at Object.__awaiter (node_modules/tslib/tslib.js:113:16)
      at MongoMemoryServer.start (node_modules/mongodb-memory-server-core/lib/MongoMemoryServer.js:78:24)
      at connect (test/database/db.js:13:21)
      at Context.<anonymous> (test/business/add.business.spec.js:19:26)
      at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**Comment by hasezoey** [comment, marked as answer]
please provide the code where you get this error

but from what i can tell already: you are trying to run something like `.start` on an already running instance (errors in 7.0)

TL;DR: use `.ensureInstance` over `.start` in this case

**Reply by iyosayi** [reply]
Okay, so here are my setup files. I basically get the error when jobs are being run in github CI, but on local they run perfectly.
```ts
import mongoose from 'mongoose'
import { MongoMemoryServer } from 'mongodb-memory-server'

const mongoServer = new MongoMemoryServer()

const opts = {
	useNewUrlParser: true,
	useUnifiedTopology: true,
	useFindAndModify: false,
}

const connect = async () => {
  await mongoServer.start()
	const mongoUri = mongoServer.getUri()
	try {
		await mongoose.connect(mongoUri, opts).then(() => console.log('Connected'))
	} catch (error) {
		console.error(error)
	}
}

const close = async () => {
	await mongoose.disconnect()
	await mongoServer.stop()
}

const clear = async () => {
	const collections = mongoose.connection.collections

	for (const key in collections) {
		await collections[key].deleteMany()
	}
}


export { connect, close, clear }
```

Then I use these exported functions in my test files like so:

```ts
import { expect } from 'chai'
import EasyGraphqlTester from 'easygraphql-tester'
import fs from 'fs'
import path from 'path'
import resolverBase from '../shared/index'
import { before, beforeEach, after } from 'mocha'
import { clear, connect, close } from '../database/db'
import userRepo from '@modules/users/infra/UserRepository'
import { createUser } from '@modules/users/use-cases'
import { makeFakeUser } from '../fixtures/user'

const schema = fs.readFileSync(
	path.join(process.cwd(), 'src', 'modules', 'schemas', 'schema.graphql'),
	{ encoding: 'utf-8' }
)

// ==== **RIGHT HERE** ====
before(async () => await connect())
beforeEach(async () => await clear())
after(async () => await close())

describe('Create User', () => {
	let tester
	before(() => {
		const resolver = {
			...resolverBase,
			Mutation: {
				createUser,
			},
		}
		tester = new EasyGraphqlTester(schema, resolver, {
			resolverValidationOptions: {
				requireResolversForResolveType: true,
			},
		})
	})

	it('Should successfully create a user', async () => {
		const mutation = `
            mutation CreateUser($data: UserInput!){
                createUser(data: $data){
                    ...on UserAuthResponse{
                        data {
                        email
                        name
                        address
                      }
                        token
                    }
                    ...on UniqueConstraintError{
                        message
                    }
                }
            }
        `

		const args = {
			data: {
				name: 'John doe',
				phoneNumber: '1234567890',
				email: 'gahhhhhhhhhhhh@gmail.com',
				password: '123567',
				address: 'earth',
			},
		}

		const {
			errors,
			data: { createUser },
		} = await tester.graphql(mutation, undefined, undefined, args)

		expect(createUser.data.name).to.be.equal('John doe')
		expect(createUser.data.address).to.equal('earth')
	})
})
```

**Reply by hasezoey** [reply]
i dont know what test runner you use (maybe mocha?), but it seems like in the CI the instance is cached (means that it is already started), and you try to call `.start` again, which will error in 7.0

fixes:
- use `.ensureInstance()` instead of `.start`
- modify your code to only call `.start` with if `if (mongoSever.state !== MongoMemoryServerStates.running)`

**Reply by iyosayi** [reply]
Thank you very much, the fixes solved the issue

**Reply by maphstay** [reply]
Only change `.start()` for `.ensureInstance()` solve my issue too. Thanks !
